# 1 Import Librerys 
Pandas and Numpy

In [1]:
import pandas as pd
import numpy as np

# 2 Read CSV lile
Review firsts lines of information, identify separator decoding type   

In [2]:
file = open('C:/Users/forqu/OneDrive/Documentos/IronHack/Github/Projetos/Shark Attack/Data/attacks.csv', 'rb')
file_lines = file.readlines(10000)
for i in range(5):
    print(file_lines[i].decode('latin-1'))
file.close()

Case Number,Date,Year,Type,Country,Area,Location,Activity,Name,Sex ,Age,Injury,Fatal (Y/N),Time,Species ,Investigator or Source,pdf,href formula,href,Case Number,Case Number,original order,,

2018.06.25,25-Jun-2018,2018,Boating,USA,California,"Oceanside, San Diego County",Paddling,Julie Wolfe,F,57,"No injury to occupant, outrigger canoe and paddle damaged",N,18h00,White shark,"R. Collier, GSAF",2018.06.25-Wolfe.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf,2018.06.25,2018.06.25,6303,,

2018.06.18,18-Jun-2018,2018,Unprovoked,USA,Georgia,"St. Simon Island, Glynn County",Standing,Adyson McNeely ,F,11,Minor injury to left thigh,N,14h00  -15h00,,"K.McMurray, TrackingSharks.com",2018.06.18-McNeely.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf,http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf,2018.06.18,2018.06.18,6302,,

201

Read CSV file with Pandas and assigne DataSet information to a variable 

In [3]:
sharkattack=pd.read_csv('C:/Users/forqu/OneDrive/Documentos/IronHack/Github/Projetos/Shark Attack/Data/attacks.csv', sep = ',', decimal = '.', encoding = 'latin-1')

# 3 Understand DataSet 
Review information, distribution, quality, quantity of NAs

In [4]:
sharkattack.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [5]:
sharkattack.shape

(25723, 24)

In [6]:
sharkattack.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

### Understand NAs distribution
understand NAs in % per columns and rows 

In [7]:
(sharkattack.isna().sum(axis = 0)/len(sharkattack)).round(2)

Case Number               0.66
Date                      0.76
Year                      0.76
Type                      0.76
Country                   0.76
Area                      0.77
Location                  0.78
Activity                  0.78
Name                      0.76
Sex                       0.78
Age                       0.87
Injury                    0.76
Fatal (Y/N)               0.78
Time                      0.89
Species                   0.87
Investigator or Source    0.76
pdf                       0.76
href formula              0.76
href                      0.76
Case Number.1             0.76
Case Number.2             0.76
original order            0.75
Unnamed: 22               1.00
Unnamed: 23               1.00
dtype: float64

In [8]:
(sharkattack.isna().sum(axis = 1)/len(sharkattack.columns)).describe().round(2)

count    25723.00
mean         0.79
std          0.36
min          0.08
25%          0.96
50%          1.00
75%          1.00
max          1.00
dtype: float64

This DataSet hav a significantnumbers of NAs, in mean 79% of values eather in rows or columns are Nas  

# 4 Cleaning DataSet
Process NAs in Columns and  Rows

### Review unnamed columns

In [9]:
sharkattack['Unnamed: 22'].value_counts()

stopped here    1
Name: Unnamed: 22, dtype: int64

In [10]:
sharkattack['Unnamed: 23'].value_counts()

Teramo             1
change filename    1
Name: Unnamed: 23, dtype: int64

### Drop Unamed columns 
As unamed columns not have only one row of information and is not usful information 

In [11]:
sharkattack = sharkattack.drop(['Unnamed: 22','Unnamed: 23'],  axis=1)

### Drop Rows 
Drop Rows with more that 20% of NAs in them  

In [12]:
sharkattack = sharkattack.dropna(axis=0, thresh= int(len(sharkattack.columns) * 0.20))
sharkattack.isna().sum() # Count NAs per Column

Case Number                  1
Date                         0
Year                         2
Type                         4
Country                     50
Area                       455
Location                   540
Activity                   544
Name                       210
Sex                        565
Age                       2831
Injury                      28
Fatal (Y/N)                539
Time                      3354
Species                   2838
Investigator or Source      17
pdf                          0
href formula                 1
href                         0
Case Number.1                0
Case Number.2                0
original order               0
dtype: int64

In [13]:
sharkattack.shape

(6302, 22)

After proccess the NAs we go form a Large Datasets fill with NAs with 25723 rows and 24 Columns to a cleaner Dataset with 6302 Rows and 22 Columns.

### Process Column Names
Treat spaces, special characters, and standarize Column Names 

In [14]:
sharkattack.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order'],
      dtype='object')

In [15]:
sharkattack.columns = [column.lower().replace(' ', '_').replace('(', '').replace(')', '').replace('.', '_') for column in sharkattack.columns]
sharkattack = sharkattack.rename(columns={"sex_": "sex", "species_": "species", "fatal_y/n" : "fatal"})
sharkattack.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal', 'time', 'species',
       'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number_1', 'case_number_2', 'original_order'],
      dtype='object')

### Treat Information 
We clean the informaion of the rows we intend to use in our analysis, columns "year", "country", "sex", "age"  and "fatal" 

### Treat Information in "age" Column
we have str contaning numbers and words, we want to have a column with numbers and str as age range. 

In [16]:
sharkattack['age'].unique()

array(['57', '11', '48', nan, '18', '52', '15', '12', '32', '10', '21',
       '34', '30', '60', '33', '29', '54', '41', '37', '56', '19', '25',
       '69', '38', '55', '35', '46', '45', '14', '40s', '28', '20', '24',
       '26', '49', '22', '7', '31', '17', '40', '13', '42', '3', '8',
       '50', '16', '82', '73', '20s', '68', '51', '39', '58', 'Teen',
       '47', '61', '65', '36', '66', '43', '60s', '9', '72', '59', '6',
       '27', '64', '23', '71', '44', '62', '63', '70', '18 months', '53',
       '30s', '50s', 'teen', '77', '74', '28 & 26', '5', '86', '18 or 20',
       '12 or 13', '46 & 34', '28, 23 & 30', 'Teens', '36 & 26',
       '8 or 10', '84', '\xa0 ', ' ', '30 or 36', '6½', '21 & ?', '75',
       '33 or 37', 'mid-30s', '23 & 20', ' 30', '7      &    31', ' 28',
       '20?', "60's", '32 & 30', '16 to 18', '87', '67', 'Elderly',
       'mid-20s', 'Ca. 33', '74 ', '45 ', '21 or 26', '20 ', '>50',
       '18 to 22', 'adult', '9 & 12', '? & 19', '9 months', '25 to 35',
  

In [17]:
def age(value): 
    """
    This Funtion take the str value, numeric or not and retur a numeric value or a categorize age range, 
    We split the str by space and compare each oth the values in the list create to identify a number to returned
    if is a not numeric string, treat the information and evaluted if the word and assigned to a especific age range or as unknown    
    
    Example 1:
        Input: "Ca. 33"
        Return: "33"
    Example 2:
        Input: "young"
        Return: "child"
    Example 3:      
        Input: "X"
        Return: "unknown"  
        
    """
    value = str(value)
    list_value = value.split(' ')
    for value in list_value :
        if value.isnumeric() : 
            try :
                value = int(value)
                return value
            except ValueError :
                value = 'unknown'
                return value
        elif ~value.isnumeric() : 
            value = value.lower().replace(' ', '').replace('(','').replace(')','')
            if value == 'toddler' or value == 'child' or value == 'teen' or value == 'adult' or value == 'middle-age' or value == 'elderly' :
                value = value
                return value
            elif value == 'teens' :
                value = 'teen'
                return value
            elif value == 'young' :
                value = 'child'
                return value
        else :
            value = 'unknown'
            return value

In [18]:
sharkattack['age']= sharkattack['age'].apply(age)
sharkattack['age']. unique()

array([57, 11, 48, None, 18, 52, 15, 12, 32, 10, 21, 34, 30, 60, 33, 29,
       54, 41, 37, 56, 19, 25, 69, 38, 55, 35, 46, 45, 14, 28, 20, 24, 26,
       49, 22, 7, 31, 17, 40, 13, 42, 3, 8, 50, 16, 82, 73, 68, 51, 39,
       58, 'teen', 47, 61, 65, 36, 66, 43, 9, 72, 59, 6, 27, 64, 23, 71,
       44, 62, 63, 70, 53, 77, 74, 5, 86, 84, 'unknown', 75, 87, 67,
       'elderly', 'adult', 1, 2, 81, 78, 'child'], dtype=object)

In [19]:
sharkattack['age'].isna().sum()

2866

We have a column with cleaner information, but we have som data as categorical range and not numeric data, so we create a new column "age_range" and categorize all the information in the column "age" as categorical age range

In [20]:
def age_rage(value) :
    """
    This funtion categorize a value in a range of age, 
    We compare the numeric value to determinated the range of the age, 
    As the value can contain a categorie, if the value is not numeric return the same values that is the categorie
    
    Example 1:
        Input: "33"
        Return: "adult"
        
    Example :
        Input: "child"
        Return: "child"
    
    """
    if type(value) == int :
        if value <= 3 :
            value = 'toddler'
            return value
        elif value > 3 and value <= 12 :
            value = 'child'
            return value
        elif value > 12 and value <= 18 :
            value = 'teen'
            return value
        elif value > 18 and value <= 35 :
            value = 'adult'
            return value
        elif value > 35 and value <= 60 :
            value = 'middle-age'
            return value
        elif value > 60 :
            value = 'elderly'
            return value
    else :
        value = value
        return value

In [21]:
sharkattack['age_range']= sharkattack['age'].apply(age_rage)
sharkattack['age_range'] = np.where(sharkattack['age_range'].isna(), 'unknown', sharkattack['age_range'])
sharkattack['age_range'].value_counts()

unknown       2868
adult         1517
teen           800
middle-age     735
child          285
elderly         89
toddler          8
Name: age_range, dtype: int64

In [22]:
sharkattack['age_range'].isna().sum()

0

We have a nwe column name "age_range", with are age categories

### Treat Information in "year" Column
we have str contaning numbers and NAs, we want to have a column with only int numbers and a min number of NAs as posible. 

In [23]:
sharkattack['year'] = np.where(sharkattack['year'].isna(), sharkattack['date'].str.slice(stop=4), sharkattack['year']) 
sharkattack['year'] = sharkattack['year'].map(int)
sharkattack['year'].unique()

ValueError: invalid literal for int() with base 10: 'Repo'

In [ ]:
def year(value):
    """
    
    """  
    date_value = str(sharkattack['date'])
    Case = str(sharkattack['case_number'])
    Case1 = str(sharkattack['case_number_1'])
    Case2 = str(sharkattack['case_number_2'])
    
    strvalue = str(value)
    if strvalue == None :
        value = int(Case.slice(stop=4))
        return value
    elif len(strvalue) < 4 : 
        if date_value.isalpha() :
            new_value = date_value
            for i in range(0,(len(new_value)-4)):
                if new_value[i].isdigit() and new_value[i+1].isdigit() and new_value[i+2].isdigit() and new_value[i+3].isdigit() :
                    value = int(new_value[i:i+3])
                    return value
        elif Case.isalpha() :
            new_value = Case
            for i in range(0,(len(new_value)-4)):
                if new_value[i].isdigit() and new_value[i+1].isdigit() and new_value[i+2].isdigit() and new_value[i+3].isdigit() :
                    value = int(new_value[i:i+3])
                    return value
        elif Case1.isalpha() :
            new_value = Case1
            for i in range(0,(len(new_value)-4)):
                if new_value[i].isdigit() and new_value[i+1].isdigit() and new_value[i+2].isdigit() and new_value[i+3].isdigit() :
                    value = int(new_value[i:i+3])
                    return value
        elif Case2.isalpha() :
            new_value = Case2
            for i in range(0,(len(new_value)-4)):
                if new_value[i].isdigit() and new_value[i+1].isdigit() and new_value[i+2].isdigit() and new_value[i+3].isdigit() :
                    value = int(new_value[i:i+3])
                    return value
    else :
        return int(value)

In [ ]:
sharkattack['year'] = sharkattack['year'].apply(year)
sharkattack['year'].unique()

In [ ]:
sharkattack[sharkattack['year'] <= 500 ].value_counts()

In [ ]:
sharkattack['year'].isna().sum()

we have all the information in this colum as a numbers

### Treat Information in "fatal" Column
The information in the column have values in Upper and Lower cases, some values do not correspond to the values expected in the column, we treat the information to have only 3 categories "y", "n" and "unknown"

In [ ]:
sharkattack['fatal'].value_counts()

In [ ]:
sharkattack['fatal'] = sharkattack['fatal'].str.replace(' ', '')
sharkattack['fatal'] = sharkattack['fatal'].str.replace('2017', 'unknown')
sharkattack['fatal'] = sharkattack['fatal'].str.replace(r'[Mm]', 'unknown')
sharkattack['fatal'] = sharkattack['fatal'].str.lower()
sharkattack['fatal'].value_counts()

In [ ]:
sharkattack['fatal'].isna().sum()

### Treat Information in "sex" Column
The information in the column have values in Upper and Lower cases, some values do not correspond to the values expected in the column, we treat the information to have only 3 categories "f", "m" and "unknown"

In [ ]:
sharkattack['sex'].unique()

In [ ]:
sharkattack['sex'] = sharkattack['sex'].str.replace(' ', '')
sharkattack['sex'] = sharkattack['sex'].str.replace('N', 'unkonwn')
sharkattack['sex'] = sharkattack['sex'].str.replace('.', 'unkonwn')
sharkattack['sex'] = sharkattack['sex'].str.replace('lli', 'unkonwn')
sharkattack['sex'] = sharkattack['sex'].str.lower()
sharkattack['sex'].value_counts()

In [ ]:
sharkattack['sex'].isna().sum()

### Treat Information in "country" Column
Treat country information for special characters, spaces and  transform all in lower cases 

In [ ]:
sharkattack['country'].unique()

In [ ]:
def country(value) :
    """
    This funtion str value and transform in lower cases, trate the special characters and spaces 
    
    Example:
        Input: "UNITED ARAB EMIRATES (UAE)"
        Return: "united_arab_emirates_uae"
    """
    if type(value) != float :
        value = str(value)
        value = value.lower().replace('/', '&').replace('(','').replace(')','').replace('?','')
        list_value = value.split(' ')
        for value in list_value :
            if value == '' or  value == 'between' or value == 'equatorial' :
                list_value.remove(value)
        value = '_'.join(list_value)
        return value
    else :
        value = 'unkonwn'
        return value

In [ ]:
sharkattack['country'] = sharkattack['country'].apply(country)
sharkattack['country'].unique()

In [ ]:
sharkattack['country'].isna().sum()

# 5  Data Analysis 


We create a new data set only with the columns we already treat the information, that are the columns usfule for owr  analys.

In [ ]:
sharkfatal = sharkattack[['country', 'year', 'fatal','age_range', 'sex']]
#sharkfatal = sharkfatal[sharkfatal['year'] != 0]
sharkfatal

In [ ]:
shark_per_country = sharkfatal.groupby('country').agg(
    first_year=pd.NamedAgg('year', 'min'),
    last_year=pd.NamedAgg('year', 'max'),
    attacks=pd.NamedAgg('fatal', 'count')) 
shark_per_country.sort_values('attacks', ascending=False).T

In [ ]:
sharkfatal = pd.DataFrame(sharkfatal)
fatalyeargroup = sharkfatal.groupby(['country','year'])['year'].count()
fatalyeargroup

In [ ]:
fatalyeargroup = pd.DataFrame(fatalyeargroup)
fatalyeargroup.sort_values('country', ascending=False)

limpar anos (trabalhar anos de 1900 para frente )

In [ ]:
fatalage = sharkfatal.groupby(['fatal','age_range'])['fatal'].count()
fatalage

In [ ]:
fatalage = pd.DataFrame(fatalage)
fatalage.sort_values('age_range')

In [ ]:
fatalyearage = sharkfatal.groupby(['fatal','age_range','year'])['fatal'].count()
fatalyearage